lib-ee/emqx_eviction_agent/src/emqx_eviction_agent_channel.erl

```diff
@@ -0,0 +1,368 @@
+%% gen_server API
+%%--------------------------------------------------------------------
+
+init([#{conninfo := OldConnInfo, clientinfo := #{clientid := ClientId} = OldClientInfo}]) ->
+    process_flag(trap_exit, true),
+    ClientInfo = clientinfo(OldClientInfo),
+    ConnInfo = conninfo(OldConnInfo),
+    case open_session(ConnInfo, ClientInfo) of
+        {ok, Channel0} ->
+            case set_expiry_timer(Channel0) of
+                {ok, Channel1} ->
+                    ?SLOG(
+                        info,
+                        #{
+                            msg => "channel_initialized",
+                            clientid => ClientId,
+                            node => node()
+                        }
+                    ),
+                    ok = emqx_cm:mark_channel_disconnected(self()),
+                    {ok, Channel1, hibernate};
+                {error, Reason} ->
+                    {stop, Reason}
+            end;
+        {error, Reason} ->
+            {stop, Reason}
+    end.
+
+handle_call(kick, _From, Channel) ->
+    {stop, kicked, ok, Channel};
+handle_call(discard, _From, Channel) ->
+    {stop, discarded, ok, Channel};
+handle_call({takeover, 'begin'}, _From, #{session := Session} = Channel) ->
+    {reply, Session, Channel#{takeover => true}};
+handle_call(
+    {takeover, 'end'},
+    _From,
+    #{
+        session := Session,
+        clientinfo := #{clientid := ClientId},
+        pendings := Pendings
+    } = Channel
+) ->
+    ok = emqx_session:takeover(Session),
+    %% TODO: Should not drain deliver here (side effect)
+    Delivers = emqx_misc:drain_deliver(),
+    AllPendings = lists:append(Delivers, Pendings),
+    ?tp(
+        debug,
+        emqx_channel_takeover_end,
+        #{clientid => ClientId}
+    ),
+    {stop, normal, AllPendings, Channel};
+handle_call(list_acl_cache, _From, Channel) ->
+    {reply, [], Channel};
+handle_call({quota, _Policy}, _From, Channel) ->
+    {reply, ok, Channel};
+handle_call(Req, _From, Channel) ->
+    ?SLOG(
+        error,
+        #{
+            msg => "unexpected_call",
+            req => Req
+        }
+    ),
+    {reply, ignored, Channel}.
+
+handle_info(Deliver = {deliver, _Topic, _Msg}, Channel) ->
+    Delivers = [Deliver | emqx_misc:drain_deliver()],
+    {noreply, handle_deliver(Delivers, Channel)};
+handle_info(expire_session, Channel) ->
+    {stop, expired, Channel};
+handle_info(Info, Channel) ->
+    ?SLOG(
+        error,
+        #{
+            msg => "unexpected_info",
+            info => Info
+        }
+    ),
+    {noreply, Channel}.
+
+handle_cast(Msg, Channel) ->
+    ?SLOG(error, #{msg => "unexpected_cast", cast => Msg}),
+    {noreply, Channel}.
+
+terminate(Reason, #{conninfo := ConnInfo, clientinfo := ClientInfo, session := Session} = Channel) ->
+    ok = cancel_expiry_timer(Channel),
+    (Reason =:= expired) andalso emqx_persistent_session:persist(ClientInfo, ConnInfo, Session),
```

Comment:
Q: should this channel publish its will message when terminating?

Reply:
`will_msg` is a channel's attribute, not session's. And this implementation does not have any (because there is no "connection" packet).